narwhal/types/src/primary.rs

```diff
@@ -85,27 +86,97 @@ impl Default for Metadata {
     }
 }
 
+// This is a versioned `Metadata` type
+// Refer to comments above the original `Metadata` struct for more details.
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq, Arbitrary, MallocSizeOf)]
+#[enum_dispatch(MetadataAPI)]
+pub enum VersionedMetadata {
+    V1(MetadataV1),
+}
+
+impl Default for VersionedMetadata {
+    fn default() -> Self {
+        // Default to the latest version
+        Self::new(ProtocolVersion::MAX)
+    }
+}
+
+impl VersionedMetadata {
+    pub fn new(_protocol_version: ProtocolVersion) -> Self {
+        Self::V1(MetadataV1 {
+            created_at: now(),
+            received_at: None,
+        })
+    }
+}
+
+#[enum_dispatch]
+pub trait MetadataAPI {
+    fn created_at(&self) -> &TimestampMs;
+    fn created_at_mut(&mut self) -> &mut TimestampMs;
+    fn received_at(&self) -> Option<TimestampMs>;
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq, Arbitrary, MallocSizeOf)]
+pub struct MetadataV1 {
+    // timestamp of when the entity created. This is generated
+    // by the node which creates the entity.
+    pub created_at: TimestampMs,
+    // timestamp of when the entity was received by the node. This will help
+    // us calculate latencies that are not affected by clock drift or network
+    // delays.
+    pub received_at: Option<TimestampMs>,
+}
+
+impl MetadataAPI for MetadataV1 {
+    fn created_at(&self) -> &TimestampMs {
+        &self.created_at
+    }
+
+    fn created_at_mut(&mut self) -> &mut TimestampMs {
+        &mut self.created_at
+    }
+
+    fn received_at(&self) -> Option<TimestampMs> {
+        self.received_at
+    }
+}
+
 #[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq, Arbitrary)]
 #[enum_dispatch(BatchAPI)]
 pub enum Batch {
     V1(BatchV1),
+    V2(BatchV2),
 }
 
-// TODO: Revisit if we should not impl Default for batch
 impl Default for Batch {
     fn default() -> Self {
-        Self::V1(BatchV1::default())
+        // Default to the latest version
+        Self::new_with_version(Vec::new(), ProtocolVersion::MAX)
     }
 }
 
 impl Batch {
     pub fn new(transactions: Vec<Transaction>) -> Self {
-        Self::V1(BatchV1::new(transactions))
+        // Use latest version
+        Batch::new_with_version(transactions, ProtocolVersion::MAX)
+    }
+
+    pub fn new_with_version(
+        transactions: Vec<Transaction>,
+        protocol_version: ProtocolVersion,
+    ) -> Self {
+        if protocol_version >= ProtocolVersion::new(11) {
```

Comment:
Instead of doing this, please use a FeatureFlag on the ProtocolConfig. We don't want raw version numbers in code. Using the config also makes testing easier.